nmagent/client.go

```diff
@@ -284,6 +288,70 @@ func (c *Client) GetHomeAz(ctx context.Context) (AzResponse, error) {
 	return homeAzResponse, nil
 }
 
+func (c *Client) RefreshSecondaryIPsIfNeeded(ctx context.Context) (refreshNeeded bool, ips []string, err error) {
+	if time.Since(c.secondaryIPLastRefreshTime) < c.secondaryIPQueryInterval {
+		return false, nil, nil
+	}
+
+	c.secondaryIPLastRefreshTime = time.Now()
+	res, err := c.getSecondaryIPs(ctx)
+	return true, res, err
+}
+
+func (c *Client) getSecondaryIPs(ctx context.Context) ([]string, error) {
+	req, err := c.buildRequest(ctx, &GetSecondaryIPsRequest{})
+	if err != nil {
+		return nil, errors.Wrap(err, "building request")
+	}
+
+	resp, err := c.httpClient.Do(req)
+	if err != nil {
+		return nil, errors.Wrap(err, "submitting request")
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, die(resp.StatusCode, resp.Header, resp.Body, req.URL.Path)
+	}
+
+	return parseSecondaryIPsFromWireServerResponse(resp)
+}
+
+func parseSecondaryIPsFromWireServerResponse(resp *http.Response) (res []string, err error) {
+	// Decode XML document.
+	var doc common.XmlDocument
+	decoder := xml.NewDecoder(resp.Body)
+	err = decoder.Decode(&doc)
+	if err != nil {
+		return nil, errors.Wrap(err, "decoding nmagent response")
+	}
+
+	// For each interface...
+	for _, i := range doc.Interface {
+		if !i.IsPrimary {
+			continue
+		}
+
+		// For each subnet on the interface...
+		for _, s := range i.IPSubnet {
+			addressCount := 0
+			// For each address in the subnet...
+			for _, a := range s.IPAddress {
+				// Primary addresses are reserved for the host.
+				if a.IsPrimary {
+					continue
+				}
+
+				res = append(res, a.Address)
+				addressCount++
+			}
+			log.Printf("Got %d addresses from subnet %s", addressCount, s.Prefix)
+		}
+	}
```

Comment:
This is all business logic that belongs in the user of the client. The responsibility of the client ends at translating the XML response to a set of structs. Also, the client only returns wrapped errors: no logging (that's the caller's responsibility).